source/Img32.pas

```diff
@@ -583,6 +583,8 @@   TImageList32 = class
 
   function MulBytes(b1, b2: Byte) : Byte;
 
+  function __Trunc(Value: Double): Integer; {$IFNDEF CPU_X86} {$IFDEF INLINE} inline; {$ENDIF} {$ENDIF}
```

Comment:
Given that __Tunc will only be called when running on CPU_X86 processors, wouldn't the following be clearer?
```
{$IFDEF CPU_X86}
  function __Trunc(Value: Double): Integer; 
{$ENDIF}
```

Reply:
OK, ignore the above question, I can see now why that doesn't work 😜.

Reply:
And retesting __Trunc() in both Delphi 7 and Delphi 11, I'm no longer seeing any difference in performance between it and System.Trunc().

Reply:
{$IFNDEF CPU_X86} should have been {$IFNDEF CPUX86}. But in debug mode the INLINE is not defined, so I overlooked it.

You are right, __Trunc is only used for the CPUX86 target. So changing it to your suggested IFDEF would work. I just didn't want to have a function that exists if you compile for one platform and doesn't exist if you compile for another platform, causing "user code" to also be IFDEF'ed.

In my tests (on an AMD Ryzen CPU Delphi 10.2 Win32) I see a performance boost of a factor of almost 2 (on bare mattel). I haven' tested it in a Virtual Machine.

### Performance with __Trunc (Delphi 10.2 Win32):
![SpeedWith_Trunc](https://github.com/AngusJohnson/Image32/assets/2139078/2c527d16-679f-46d1-95f2-7af440b01de3)

### Performance with System.Trunc (Delphi 10.2 Win32):
![SpeedWithOut_Trunc](https://github.com/AngusJohnson/Image32/assets/2139078/0bb2cd7a-9623-4e52-ab09-f0d052d875a3)

### There is an even faster Trunc version possible...
... but not that much faster than your __Trunc function.
With SSE3, Intel introduced a new FPU instruction that is the implementation for TRUNC. And by patching TRUNC to use that instruction, all application code would benefit from the "not having to change the FPU ControlWord". But that would require "online code patching" (not as wild as it sounds), some assembler and SSE3 detection (unless CPUs after 2004/2005 are the minimum requirement). Using this technique I got another 100ms faster in a 50 runs benchmark. But it is nothing FPC could use, involves assembler and it uses code patching, that a library souldn't do.

Reply:
Testing only __Trunc vs System.Trunc in a tight loop may not show the performance difference.

- __Trunc uses the "register" calling convention where "double" types are passed over the stack.
- System.Trunc uses a special calling convention where the "double" parameter is passed in the st(0) FPU register.

Now it depends on what your code does. BTW: The compiler optimizes "System.Trunc(10.2)" away and replaces it with 10.


### Calculation result written to a variable before calling Trunc
```
d := 10.2 * x;
__Trunc(d);
System.Trunc(d);
```
Here __Trunc wins, because System.Trunc has to load the memory variable d into the st(0) FPU register first and has the x87 ControlWord overhead.

### Calculation done as the actual parameter for Trunc
```
__Trunc(10.2 * x);
System.Trunc(10.2 * x);
```
Here System.Trunc may win (almost not measurably). System.Trunc has the advantage, that the calculation's result is already in st(0). It depends on how much the x87 ControlWord save/change/restore costs.
__Trunc must first read st(0) and push it onto the stack.


### Changing the x87 ControlWord:
> FPCW: An instruction that writes to the floating-point control word, such as FLDCW, may stall the pipeline until all earlier uops complete execution. The conditions depend on the microarchitecture and the FPCW bits that are modified (see Section 3.8.3 of the Intel optimization manual).
(Source: https://stackoverflow.com/questions/60264057/what-resource-stall-other-might-mean)

Reply:
In the Image32 performance testing above, I don't understand why the 50x loop test performance (for drawing) averages more that 150x slower than the single instance performance.

Reply:
The single instance performance is for the little image on the right side. The "Benchmark" button uses the client area and increases the size from small to large. So more pixels have to be rendered. It also doesn't release memory. For a very large SVGs with (90 MB), I ran out of memory with more than 2 loops with Win32.

![Benchmark](https://github.com/AngusJohnson/Image32/assets/2139078/7e0b5fc3-c622-4947-a851-1c092ef84da8)

Reply:
> . For a very large SVGs with (90 MB), I ran out of memory with more than 2 loops with Win32.

I'm aware that my SVG code isn't well written. It started out as a quick hack for something I want intending to contribute to the library. And it's grown now to the point that rewriting it will be a very big job and I'm really occupied with another project. Nevertheless I'm certainly appreciating your most helpful code corrections and contributions.

Reply:
Well, it's the Benchmark project that doesn't release the memory. It keeps all images in a SVGImageList. There is no memory leak in Image32, that I know of.

The 90 MB SVG I'm referring to here, is [Koppen-Geiger_Map_Cfc_present.svg](https://upload.wikimedia.org/wikipedia/commons/f/fb/Koppen-Geiger_Map_Cfc_present.svg) (Warning: Even Chrome/Skia needs time to render it).

Reply:
> There is no memory leak in Image32, that I know of.

Phew! 🤣